src/lib.rs

```diff
@@ -1,7 +1,372 @@
+            f,
+            "Value is bigger than the configured max size of the cache"
+        )
+    }
+}
+
+/// A cache that holds a limited number of key-value pairs. When the capacity of
+/// the cache is exceeded, the least-recently-used (where "used" means a look-up
+/// or putting the pair into the cache) pair is automatically removed.
+pub struct WeightCache<K, V, S = hash_map::RandomState> {
+    max: usize,
+    current: usize,
+    inner: LinkedHashMap<K, V, S>,
+}
+impl<K, V, S> fmt::Debug for WeightCache<K, V, S> {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("WeightCache")
+            .field("max", &self.max)
+            .field("current", &self.current)
+            .finish()
+    }
+}
+
+impl<K: Hash + Eq, V: Weigheable> Default for WeightCache<K, V> {
+    fn default() -> Self {
+        WeightCache::<K, V, RandomState>::new(NonZeroUsize::new(usize::MAX).expect("MAX > 0"))
+    }
+}
+
+impl<K: Hash + Eq, V: Weigheable> WeightCache<K, V> {
+    pub fn new(capacity: NonZeroUsize) -> Self {
+        Self {
+            max: capacity.get(),
+            current: 0,
+            inner: LinkedHashMap::new(),
+        }
+    }
+}
+impl<K: Hash + Eq, V: Weigheable, S: BuildHasher> WeightCache<K, V, S> {
+    /// Create a [`WeightCache`] with a custom hasher.
+    pub fn with_hasher(capacity: NonZeroUsize, hasher: S) -> Self {
+        Self {
+            max: capacity.get(),
+            current: 0,
+            inner: LinkedHashMap::with_hasher(hasher),
+        }
+    }
+
+    /// Returns a reference to the value corresponding to the given key, if it
+    /// exists.
+    pub fn get(&mut self, k: &K) -> Option<&V> {
+        self.inner.get_refresh(k).map(|v| v as &V)
+    }
+
+    /// Returns a mutable reference to the value corresponding to the given key,
+    /// if it exists.
+    pub fn get_mut(&mut self, k: &K) -> Option<&mut V> {
+        self.inner.get_refresh(k)
+    }
+
+    /// Returns the number of key-value pairs in the cache.
+    pub fn len(&self) -> usize {
+        self.inner.len()
+    }
+
+    /// Returns `true` if the cache contains no key-value pairs.
+    pub fn is_empty(&self) -> bool {
+        self.inner.is_empty()
+    }
+
+    /// Inserts a key-value pair into the cache. Returns an error if the value is
+    /// too bigger than the cache's configured max size.
+    pub fn put(&mut self, key: K, value: V) -> Result<(), ValueTooBigError> {
+        let weight = V::measure(&value);
+        if weight > self.max {
+            Err(ValueTooBigError)
+        } else {
+            self.current += weight;
+            // did we remove an element?
+            if let Some(x) = self.inner.insert(key, value) {
+                self.current -= V::measure(&x);
+            }
+
+            while self.current > self.max && !self.inner.is_empty() {
+                if let Some((_, v)) = self.inner.pop_front() {
+                    self.current -= V::measure(&v);
```

Comment:
So the measure of an item is not cached. In that case we need to document that measure needs to be a pure function, otherwise things will get very strange...

Reply:
I just went for KISS; but just adding it to the payload should be straight-forward. Will do.

Reply:
this does the same as `shrink_to_fit`, why do we `shrink_to_fit` twice?

Reply:
Right, oversight.

Reply:
I added the size to the cached value.